api/src/main/java/org/apache/iceberg/metrics/ScanReport.java

```diff
@@ -0,0 +1,370 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.metrics;
+
+import java.io.Serializable;
+import java.time.Duration;
+import java.util.concurrent.TimeUnit;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.expressions.Expression;
+import org.apache.iceberg.metrics.MetricsContext.Counter;
+import org.apache.iceberg.relocated.com.google.common.base.MoreObjects;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+
+/**
+ * A Table Scan report that contains all relevant information from a Table Scan.
+ */
+public class ScanReport implements Serializable {
+
+  private final String tableName;
+  private final long snapshotId;
+  private final Expression filter;
+  private final Schema projection;
+  private final int resultDataFiles;
+  private final int totalDataManifests;
+  private final int totalDeleteManifests;
+  private final int scannedDataManifests;
+  private final int scannedDeleteManifests;
+  private final int totalDataFiles;
+  private final int totalDeleteFiles;
+  private final Duration totalPlanningDuration;
+  private final long totalFileSizeInBytes;
+  private final int indexedDeleteFiles;
+  private final int resultDeleteFiles;
+  private final int skippedDataManifests;
+  private final int skippedDeleteManifests;
+  private final int skippedDataFiles;
+  private final int skippedDeleteFiles;
+
+  private ScanReport(
+      String tableName, long snapshotId, Expression filter, Schema projection,
+      int resultDataFiles, int totalDataManifests, int totalDeleteManifests, int scannedDataManifests,
+      int scannedDeleteManifests, int totalDataFiles, int totalDeleteFiles, Duration totalPlanningDuration,
+      long totalFileSizeInBytes, int indexedDeleteFiles, int resultDeleteFiles,
+      int skippedDataManifests, int skippedDeleteManifests,
+      int skippedDataFiles, int skippedDeleteFiles) {
+    this.tableName = tableName;
+    this.snapshotId = snapshotId;
+    this.filter = filter;
+    this.projection = projection;
+    this.resultDataFiles = resultDataFiles;
```

Comment:
I had a conversation with Ryan about what to do with ScanReport and ScanMetrics.  I think we actually want to remove the final values from the scan report here.  I get that will make it easy to serialize a pojo result, but there's a separation of concerns that I think we can achieve rather easily.

Rather than collect all of the metric results in the ScanReport, just keep a reference to the ScanMetrics with the actual counters/timers.  This will allow the ScanReporter to manage the structure of the response.

For example:

```java
ScanReport:
  - table identifier
  - snapshot id
  - filter
  - projection
  - ScanMetrics:
      - <various metrics>
```

The scan reporter can then construct the payload dynamically based off of the Scan report (espcially if we make a way to iterate over the metrics):

```json
{
  "table-identifier": "a.b.c",
  "snapshot-id": 123,
  "filter": "foo=bar",
  "projection": "c1, c2",
  "metrics": {
    "<name>": {
      "type": "counter",   # we probably type metric type info and 
      "value": "val"
    },
    "<name>": {
      "type": "timer",
      "value": "val",
      "histogram": { "p50":  213, "p99":  18983 },
      "min": 23,
      "max": 20199
    }
  }
}
```

This allows us to separate the way we track metrics and the way we construct the payload so that if a future metric (say from micrometer) has a histogram we could then just update the reporter without have to create a new object structure for the values in ScanReport.  As you can see from what the json might look like, the metric info can get more complicated than we probably want to track here.

Reply:
I like that idea, updated the code to reflect that